### ISSUE TITLE
Initializing the SDL Backend causes OOM error.

### ISSUE THREAD

**Issue by JosephABudd** [david-vanderson/dvui#74]
**I am really excited about how dvui has developed. The dvui team is taking zig to another level.**

I'm not vendoring dvui any more. I'm using build.zig.zon setup with zig fetch.

The dvui commit https://github.com/david-vanderson/dvui/commit/27b59c5f25350ad4481110eecd0920b828e61a30 was working great for me once I upgraded my app's zig code to 0.12.0.

So now I'm using the current commit which is https://github.com/david-vanderson/dvui/commit/8645f753c1b2eb93f1bd795e2d24469f5493b670. My app's code is upgraded to zig 0.13.0 and builds without errors using zig 0.13.0.

When I run my app I get this out of memory error. Any ideas?

```log
error: OutOfMemory
/usr/local/zig/lib/std/mem/Allocator.zig:225:89: 0x176305d in allocBytesWithAlignment__anon_30550 (crud)
    const byte_ptr = self.rawAlloc(byte_count, log2a(alignment), return_address) orelse return Error.OutOfMemory;
                                                                                        ^
/usr/local/zig/lib/std/mem/Allocator.zig:211:5: 0x16f9f56 in allocWithSizeAndAlignment__anon_27393 (crud)
    return self.allocBytesWithAlignment(alignment, byte_count, return_address);
    ^
/usr/local/zig/lib/std/mem/Allocator.zig:129:5: 0x1611788 in alloc__anon_24357 (crud)
    return self.allocAdvancedWithRetAddr(T, null, n, @returnAddress());
    ^
/usr/local/zig/lib/std/mem/Allocator.zig:326:21: 0x1759c74 in dupeZ__anon_30122 (crud)
    const new_buf = try allocator.alloc(T, m.len + 1);
                    ^
/usr/local/zig/lib/std/process.zig:1939:23: 0x175b2cd in createEnvironFromExisting (crud)
        envp_buf[i] = try arena.dupeZ(u8, mem.span(line));
                      ^
/usr/local/zig/lib/std/process/Child.zig:622:23: 0x16f1bb8 in spawnPosix (crud)
            break :m (try process.createEnvironFromExisting(arena, std.c.environ, .{
                      ^
/usr/local/zig/lib/std/process/Child.zig:242:9: 0x160cb78 in spawn (crud)
        return self.spawnPosix();
        ^
/home/nil/.cache/zig/p/1220f35274f3c12ac2dfc02edaa9c815cdcdfc657e564d74541aa9c3502a251d67c0/src/backends/SDLBackend.zig:138:21: 0x160b28a in init (crud)
                    try child.spawn();
                    ^
/home/nil/zig/crud/src/main.zig:42:23: 0x16e7278 in main (crud)
    var sdl_backend = try SDLBackend.init(.{
                      ^
```

**Comment by iacore** [comment]
> Any ideas?

no idea. Can you share your code?

**Comment by JosephABudd** [comment]
# blow are 2 files.

1. build.zig.zon
2. main.zig

## build.zig.zon
```
.{
    .name = "crud",
    .version = "0.0.0",
    .dependencies = .{
        .fridge = .{
            .path = "./vendor/fridge",
        },
        .dvui = .{
            .url = "https://github.com/david-vanderson/dvui/archive/8645f753c1b2eb93f1bd795e2d24469f5493b670.tar.gz",
            .hash = "12202a61da84074990523e3eca0c1a76507362f4d7354e5f884ae989a504372ab0ff",
        },
    },
    // Specifies the set of files and directories that are included in this package.
    // Only files and directories listed here are included in the `hash` that
    // is computed for this package.
    // Paths are relative to the build root. Use the empty string (`""`) to refer to
    // the build root itself.
    // A directory listed here means that all files within, recursively, are included.
    .paths = .{
        // This makes *all* files, recursively, included in this package. It is generally
        // better to explicitly list the files and directories instead, to insure that
        // fetching from tarballs, file system paths, and version control all result
        // in the same contents hash.
        "",
        // For example...
        //"build.zig",
        //"build.zig.zon",
        //"src",
        //"LICENSE",
        //"README.md",
    },
}
```

## main.zig

This is the same main.zig that I used for the previous version of dvui which used zig version 0.12.0.

```
/// This is where the application starts.
/// This file was generated by kickzig when you created the framework.
/// This file will be never be touched by kickzig.
/// You are free to edit this file.
const std = @import("std");
const dvui = @import("dvui");
const SDLBackend = @import("SDLBackend");

const _backend_ = @import("backend/api.zig");
const _channel_ = @import("channel");
const _closer_ = @import("closer");
const _closedownjobs_ = @import("closedownjobs");
const _frontend_ = @import("frontend/api.zig");
const _modal_params_ = @import("modal_params");
const _startup_ = @import("startup");
const ExitFn = @import("various").ExitFn;
const MainView = @import("framers").MainView;

const window_icon_png = @embedFile("zig-favicon.png");

// KICKZIG TODO:
// When the user clicks the window's X:
// - If a modal screen is not shown:
//   * then the app will just close.
// - If a modal screen is shown:
//   * If force_close == true: then the app will just close.
//   * If force_close == false: then the app will not close until after the modal screen is hidden (closes).
const force_close: bool = true;

// General Purpose Allocator for frontend-state, backend and channels.
var gpa_instance = std.heap.GeneralPurposeAllocator(.{}){};
const gpa = gpa_instance.allocator();

const vsync = true;
var show_dialog_outside_frame: bool = false;

/// This example shows how to use the dvui for a normal application:
/// - dvui renders the whole application
/// - render frames only when needed
pub fn main() !void {
    // init SDL sdl_backend (creates OS window)
    var sdl_backend = try SDLBackend.init(.{
        .size = .{ .w = 500.0, .h = 400.0 },
        .min_size = .{ .w = 500.0, .h = 400.0 },
        .vsync = vsync,
        .title = "crud",
        .icon = window_icon_png, // can also call setIconFromFileContent()
    });
    defer sdl_backend.deinit();

    // init dvui Window (maps onto a single OS window)
    var win = try dvui.Window.init(@src(), 0, gpa, sdl_backend.backend());
    // win.content_scale = sdl_backend.initial_scale * 1.5;
    defer win.deinit();

    var main_view: *MainView = undefined;
    var close_down_jobs: *_closedownjobs_.Jobs = try _closedownjobs_.Jobs.init(gpa);
    defer close_down_jobs.deinit();
    const exit: ExitFn = try _closer_.init(gpa, close_down_jobs, &win);
    defer _closer_.deinit();

    // The channels between the front and back ends.
    const back_to_front_channels: *_channel_.BackendToFrontend = try _channel_.BackendToFrontend.init(gpa, exit);
    defer back_to_front_channels.deinit();
    const front_to_back_channels: *_channel_.FrontendToBackend = try _channel_.FrontendToBackend.init(gpa, exit);
    defer front_to_back_channels.deinit();
    const triggers: *_channel_.Trigger = try _channel_.Trigger.init(gpa, exit);
    defer triggers.deinit();

    // Initialize the front end.
    // See src/deps/startup/api.zig
    var startup_frontend: _startup_.Frontend = _startup_.Frontend{
        .allocator = gpa,
        .window = &win,
        .send_channels = front_to_back_channels,
        .receive_channels = back_to_front_channels,
        .main_view = undefined,
        .close_down_jobs = close_down_jobs,
        .exit = exit,
        .screen_pointers = undefined,
    };
    main_view = try MainView.init(startup_frontend);
    defer main_view.deinit();
    startup_frontend.setMainView(main_view);
    try _frontend_.init(&startup_frontend);
    defer _frontend_.deinit();
    _closer_.set_screens(main_view);

    // Initialize and kick-start the back end.
    try _backend_.init(
        .{
            .allocator = gpa,
            .send_channels = back_to_front_channels,
            .receive_channels = front_to_back_channels,
            .close_down_jobs = close_down_jobs,
            .triggers = triggers,
            .exit = exit,
        },
    );
    defer _backend_.deinit();

    // KICKZIG TODO: See backend.kickStart();
    try _backend_.kickStart();

    var theme_set: bool = false;
    main_loop: while (true) {

        // beginWait coordinates with waitTime below to run frames only when needed
        const nstime = win.beginWait(sdl_backend.hasEvent());

        // marks the beginning of a frame for dvui, can call dvui functions after this
        try win.begin(nstime);

        // set the theme.
        if (!theme_set) {
            theme_set = true;
            const dark_theme = &dvui.Adwaita.dark;
            dvui.themeSet(dark_theme);
        }

        // send all SDL events to dvui for processing
        const quit = try sdl_backend.addAllEvents(&win);
        // The state of the app's closer.
        switch (_closer_.context()) {
            .none => {
                if (quit) {
                    // User clicked window's X to close the window.
                    // Start the closing process.
                    _closer_.close("Bye-bye.", force_close);
                } else {
                    // Not closing.
                    // Continue running the app.
                }
            },
            .forced => {
                // The previous frame set closer_state to .forced.
                // So on this frame, force the close.
                // Start the closing process.
                _closer_.forced();
            },
            .started => {
                // The close process has already started.
                // The close process is now running.
            },
            .completed => {
                // The close process has completed.
                // Stop framing.
                break :main_loop;
            },
            .waiting => {
                // The close process is now waiting for the current modal screen to close.
                // When that modal screen closes, the closing state will switch to .started.
                _closer_.waiting();
            },
        }

        // if dvui widgets might not cover the whole window, then need to clear
        // the previous frame's render
        sdl_backend.clear();

        {
            // Frame the front-end.
            var arena_allocator = std.heap.ArenaAllocator.init(std.heap.page_allocator);
            defer arena_allocator.deinit();
            try _frontend_.frame(arena_allocator.allocator());
        }

        // marks end of dvui frame, don't call dvui functions after this
        // - sends all dvui stuff to backend for rendering, must be called before renderPresent()
        const end_micros = try win.end(.{});

        // cursor management
        sdl_backend.setCursor(win.cursorRequested());

        // render frame to OS
        sdl_backend.renderPresent();

        // waitTime and beginWait combine to achieve variable framerates
        const wait_event_micros = win.waitTime(end_micros, null);
        sdl_backend.waitEventTimeout(wait_event_micros);

        // Example of how to show a dialog from another thread (outside of win.begin/win.end)
        if (show_dialog_outside_frame) {
            show_dialog_outside_frame = false;
            try dvui.dialog(@src(), .{ .window = &win, .modal = false, .title = "Dialog from Outside", .message = "This is a non modal dialog that was created outside win.begin()/win.end(), usually from another thread." });
        }
    }
}

```

**Comment by JosephABudd** [comment]
Done and also included my vendor folder.
https://github.com/JosephABudd/crud

Zig seems to have some OOM issues: https://github.com/ziglang/zig/issues?q=is%3Aissue+is%3Aopen+OutOfMemory

**Comment by JosephABudd** [comment]
Re: my enthusiasm.
It's sincere. I really like dvui.
I actually wrote that after looking at the web example that is linked to. I didn't realize it would look that nice.

**Comment by iacore** [comment]
> Re: my enthusiasm.
It's sincere. I really like dvui.

... still, [i can't handle enthusiasm.](https://www.1a-insec.net/frag/18-zig-what/)

> I actually wrote that after looking at the web example that is linked to. I didn't realize it would look that nice.

you should thank @david-vanderson for this.

***

btw, your repo is still not reproducible. it's missing `gitmodules`.

**Comment by JosephABudd** [comment]
I don't even have a .gitmodules in my crud folder. Probably because I copied the vendor folder from 0.12.0 version of my crud.

So I copied it over and pushejd it to the repo.

**Comment by iacore** [comment]
ok. this seems like a dvui bug. thanks for the bug report.

**Comment by JosephABudd** [comment]
Thanks, have a good one.
Liked the article.

**Comment by iacore** [comment]
Fixed in dc33404.

**Comment by david-vanderson** [comment]
@JosephABudd Thanks for the generous comments!  It's very nice to hear that it's working well for you.  Thanks for the bug report!

@iacore Sorry to hear about the zig drama.

Looking at the issue here, it seems like my estimate for how much memory needed for looking up those environment variables was too low.  Thanks so much @iacore for the investigation and quick fix!
